### PR TITLE
creating and selecting stake transactions [proposal and wip]

### PR DESCRIPTION
**Pull request by alajko**

### CODE REVIEW

File: internal/utils/utils.go [hunk 1]
```diff
@@ -13,15 +13,26 @@ import (
...
 	"github.com/harmony-one/harmony/p2p"
 )
 
-var lock sync.Mutex
+var (
+	lock     sync.Mutex
+	key, err = ecdsa.GenerateKey(eth_crypto.S256(), strings.NewReader("Beaconchain Dummy Address"))
+	//BeaconChainAddress is available across all packages. Make it constant.
+	BeaconChainAddress = eth_crypto.PubkeyToAddress(key.PublicKey)
+)
```
Comment (alajko) [inline internal/utils/utils.go]: I should make this a constant.

File: core/types/transaction.go [hunk 2]
```diff
@@ -84,6 +85,12 @@ func NewContractCreation(nonce uint64, shardID uint32, amount *big.Int, gasLimit
 	return newTransaction(nonce, nil, shardID, amount, gasLimit, gasPrice, data)
 }
 
+// NewStakeTransaction constructs a stake transaction.
+// Currently a stake transaction is sent to a constant beaconchain address and the shard id is 0.
```
Comment (rlan35) [inline core/types/transaction.go]: The second line of comment better be somewhere else where you send the transaction. The stake transaction by itself can be implemented without hack, so no need to put how the stake tx is used in this package.

File: core/types/transaction.go [hunk 3]
```diff
@@ -84,6 +85,12 @@ func NewContractCreation(nonce uint64, shardID uint32, amount *big.Int, gasLimit
...
+func NewStakeTransaction(nonce uint64, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
```
[the remark this answers is not in the document]
Comment (alajko) [reply core/types/transaction.go]: Thanks. That makes more sense, `staking` could be one of the services @mikedoan can append on service manager as well.

File: node/worker/worker.go [hunk 4]
```diff
@@ -41,6 +42,37 @@ type Worker struct {
...
+func (w *Worker) SelectStakeTransactionsForNewBlock(txs types.Transactions, maxNumTxs int) (types.Transactions, types.Transactions, types.Transactions) {
+	if w.current.gasPool == nil {
+		w.current.gasPool = new(core.GasPool).AddGas(w.current.header.GasLimit)
+	}
+	selected := types.Transactions{}
+	unselected := types.Transactions{}
+	invalid := types.Transactions{}
+	for _, tx := range txs {
+		if tx.ShardID() == 0 && *tx.Recipient() == utils.BeaconChainAddress {
+			selected = append(selected, tx)
```
[the remark this answers is not in the document]
Comment (alajko) [reply node/worker/worker.go]: Thanks, good catch.

File: cmd/client/txgen/txgen/account_txs_generator.go [hunk 5]
```diff
@@ -18,6 +18,22 @@ type Settings struct {
 	CrossShardRatio   int
 }
 
+// GenerateSimulatedStakeTransactionsAccount generates simulated stake transaction for account model.
```
[the remark this answers is not in the document]
Comment (alajko) [reply cmd/client/txgen/txgen/account_txs_generator.go]: This was added to demonstrate how we can test staking of identities. New ones, should only be added one at a time as you mention.

File: core/types/transaction.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rlan35) [reply core/types/transaction.go]: Yea, it can be part of a "mining"/"validating" service.

**Pull request closed without merging**